opts/opts.go

```diff
@@ -486,3 +490,59 @@ func (m *MemSwapBytes) UnmarshalJSON(s []byte) error {
 	b := MemBytes(*m)
 	return b.UnmarshalJSON(s)
 }
+
+// GenericResource is a concept that a user can use to advertise user-defined
+// resources on a node and thus better place services based on these resources.
+// E.g: NVIDIA GPUs, Intel FPGAs, ...
+// See https://github.com/docker/swarmkit/blob/master/design/generic_resources.md
+type GenericResource []swarm.GenericResource
+
+// ValidateSingleGenericResource validates that a single entry in the
+// generic resource list is valid.
+// i.e 'GPU=UID1' is valid however 'GPU:UID1' or 'UID1' isn't
+func ValidateSingleGenericResource(val string) (string, error) {
+	if strings.Count(val, "=") < 1 {
+		return "", fmt.Errorf("invalid generic-resource format `%s` expected `name=value`", val)
+	}
+
+	return val, nil
+}
+
+// ParseGenericResources parses an array of Generic resourceResources
+func ParseGenericResources(value []string) ([]swarm.GenericResource, error) {
+	if len(value) == 0 {
+		return nil, nil
+	}
+
+	resources, err := genericresource.Parse(value)
```

Comment:
It looks like `ValidateSingleGenericResource()` is doing partial validation and the rest is handled by `ParseGenericResources()` when we go to convert it into a swarm type.

I think there are a couple minor issues with this. One is that the errors returned by `ParseGenericResources()` don't really include enough context, so it'll be hard for a user to fix the issue.

I think the ideal solution to this would be a custom flag Opt type (instead of the generic ListOpt) however that wouldn't let us reuse any of `ParseGenericResources` because it requires the entire list as input, instead of a single key=value pair.

So maybe the next best option is to just wrap any errors returned by `genericresource.Parse()` with more context about the error. Something like:

```go
errors.Wrapf(err, "invalid generic resource specification")
```